privval/signer_validator_endpoint.go

```diff
@@ -0,0 +1,230 @@
+type SignerValidatorEndpoint struct {
+	cmn.BaseService
+
+	listener net.Listener
+
+	// ping
+	cancelPingCh    chan struct{}
+	pingTicker      *time.Ticker
+	heartbeatPeriod time.Duration
+
+	// signer is mutable since it can be reset if the connection fails.
+	// failures are detected by a background ping routine.
+	// All messages are request/response, so we hold the mutex
+	// so only one request/response pair can happen at a time.
+	// Methods on the underlying net.Conn itself are already goroutine safe.
+	mtx sync.Mutex
+
+	// TODO: Signer should encapsulate and hide the endpoint completely. Invert the relation
+	signer *SignerRemote
+}
+
+// Check that SignerValidatorEndpoint implements PrivValidator.
+var _ types.PrivValidator = (*SignerValidatorEndpoint)(nil)
+
+// NewSignerValidatorEndpoint returns an instance of SignerValidatorEndpoint.
+func NewSignerValidatorEndpoint(logger log.Logger, listener net.Listener) *SignerValidatorEndpoint {
+	sc := &SignerValidatorEndpoint{
+		listener:        listener,
+		heartbeatPeriod: heartbeatPeriod,
+	}
+
+	sc.BaseService = *cmn.NewBaseService(logger, "SignerValidatorEndpoint", sc)
+
+	return sc
+}
+
+//--------------------------------------------------------
+// Implement PrivValidator
+
+// GetPubKey implements PrivValidator.
+func (ve *SignerValidatorEndpoint) GetPubKey() crypto.PubKey {
+	ve.mtx.Lock()
+	defer ve.mtx.Unlock()
+	return ve.signer.GetPubKey()
+}
+
+// SignVote implements PrivValidator.
+func (ve *SignerValidatorEndpoint) SignVote(chainID string, vote *types.Vote) error {
+	ve.mtx.Lock()
+	defer ve.mtx.Unlock()
+	return ve.signer.SignVote(chainID, vote)
+}
+
+// SignProposal implements PrivValidator.
+func (ve *SignerValidatorEndpoint) SignProposal(chainID string, proposal *types.Proposal) error {
+	ve.mtx.Lock()
+	defer ve.mtx.Unlock()
+	return ve.signer.SignProposal(chainID, proposal)
+}
+
+//--------------------------------------------------------
+// More thread safe methods proxied to the signer
+
+// Ping is used to check connection health.
+func (ve *SignerValidatorEndpoint) Ping() error {
+	ve.mtx.Lock()
+	defer ve.mtx.Unlock()
+	return ve.signer.Ping()
+}
+
+// Close closes the underlying net.Conn.
+func (ve *SignerValidatorEndpoint) Close() {
+	ve.mtx.Lock()
+	defer ve.mtx.Unlock()
+	if ve.signer != nil {
+		if err := ve.signer.Close(); err != nil {
+			ve.Logger.Error("OnStop", "err", err)
+		}
+	}
+
+	if ve.listener != nil {
+		if err := ve.listener.Close(); err != nil {
+			ve.Logger.Error("OnStop", "err", err)
+		}
+	}
+}
+
+//--------------------------------------------------------
+// Service start and stop
+
+// OnStart implements cmn.Service.
+func (ve *SignerValidatorEndpoint) OnStart() error {
+	if closed, err := ve.reset(); err != nil {
+		ve.Logger.Error("OnStart", "err", err)
+		return err
+	} else if closed {
+		return fmt.Errorf("listener is closed")
+	}
+
+	// Start a routine to keep the connection alive
+	ve.cancelPingCh = make(chan struct{}, 1)
```

Comment:
why capacity == 1?

Reply:
not sure, implementation was like that before.
Remember anyway that the next PR will remove the Heartbeat/Pings. so I would say this is not very important.